File: Solutions/C++/Accepted/543-diameter-of-binary-tree.cpp
/*
Problem Title: 543 - Diameter of Binary Tree
Lang: C++
Runtime: 16 ms
Memory: 17.4 MB
Problem Url: https://leetcode.com/problems/diameter-of-binary-tree
Submission Url: https://leetcode.com/problems/diameter-of-binary-tree/submissions/322972707
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    
    int helper(TreeNode* root, int& ans){
        if(root == NULL){
            return 0;
        }
        int left = helper(root->left, ans);
        int right = helper(root->right, ans);
        ans = max({ans, left+right});
        return max(left,right)+1;
    }
    
    int diameterOfBinaryTree(TreeNode* root) {
        int ans = 0;
        helper(root, ans);
        return ans;
    }
};

File: Solutions/C#/Accepted/543-diameter-of-binary-tree.cs
/*
Problem Title: 543 - Diameter of Binary Tree
Lang: C#
Runtime: 84 ms
Memory: 25.5 MB
Problem Url: https://leetcode.com/problems/diameter-of-binary-tree
Submission Url: https://leetcode.com/problems/diameter-of-binary-tree/submissions/305654208
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int DiameterOfBinaryTree(TreeNode root) {
        int ans = 0;
        DiameterOfBinaryTreeHelper(root, ref ans);
        return ans;
    }
    
    private int DiameterOfBinaryTreeHelper(TreeNode root, ref int ans){
        if(root == null) {
            return 0;
        }
        int left_height = DiameterOfBinaryTreeHelper(root.left, ref ans);
        int right_height = DiameterOfBinaryTreeHelper(root.right, ref ans);
        ans = Math.Max(ans, left_height + right_height);
        return Math.Max(left_height, right_height) + 1;
    }
}